### PR TITLE
refactor(dialect): refactor dialect reserve words

### PR DESCRIPTION
**Pull request by jesse23**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Refactor dialect reserved words

### CODE REVIEW

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -35,6 +34,9 @@ class QueryGenerator {
     // dialect name
     this.dialect = options._dialect.name;
     this._dialect = options._dialect;
+
+    // wrap quoteIdentifier with common logic
+    this._updateQuoteIdentifier();
```
Comment (sdepold) [inline lib/dialects/abstract/query-generator.js:39]: Can we call it `_initQuoteIdentifier`?

File: lib/dialects/postgres/query-generator.js [hunk 2]
```diff
@@ -900,6 +906,36 @@ class PostgresQueryGenerator extends AbstractQueryGenerator {
...
+   * Quote identifier in sql clause
+   *
+   * @param {string} identifier
+   * @param {boolean} force
+   *
+   * @returns {string}
+   */
+  quoteIdentifier(identifier, force) {
+    const optForceQuote = force || false;
+    const optQuoteIdentifiers = this.options.quoteIdentifiers === false ? false : true;
```
Comment (sdepold) [inline lib/dialects/postgres/query-generator.js:920]: I tend to believe that this is the same as `this.options.quoteIdentifiers !== false`

File: lib/dialects/snowflake/query-generator.js [hunk 3]
```diff
@@ -25,6 +25,14 @@ const FOREIGN_KEY_FIELDS = [
   'REFERENCED_COLUMN_NAME as referencedColumnName'
 ].join(',');
 
+/**
+ * list of reserved words in Snowflake
+ * source: https://docs.snowflake.com/en/sql-reference/reserved-keywords.html
+ */
+const SNOWFLAKE_RESERVED_WORDS = 'account,all,alter,and,any,as,between,by,case,cast,check,column,connect,connections,constraint,create,cross,current,current_date,current_time,current_timestamp,current_user,database,delete,distinct,drop,else,exists,false,following,for,from,full,grant,group,gscluster,having,ilike,in,increment,inner,insert,intersect,into,is,issue,join,lateral,left,like,localtime,localtimestamp,minus,natural,not,null,of,on,or,order,organization,qualify,regexp,revoke,right,rlike,row,rows,sample,schema,select,set,some,start,table,tablesample,then,to,trigger,true,try_cast,union,unique,update,using,values,view,when,whenever,where,with'.split(',');
+  
+
```
Comment (sdepold) [inline lib/dialects/snowflake/query-generator.js:34]: Can you remove the empty lines?

File: lib/dialects/snowflake/query-generator.js [hunk 4]
```diff
@@ -628,19 +636,49 @@ class SnowflakeQueryGenerator extends AbstractQueryGenerator {
     let fragment = '';
 
     /* eslint-disable */
-    if (options.offset != null && options.limit == null) {
+    /*if (options.offset != null && options.limit == null) {
```
Comment (sdepold) [inline lib/dialects/snowflake/query-generator.js:639]: What's going on here ::)

File: package.json [hunk 5]
```diff
@@ -207,6 +207,7 @@
     "----------------------------------------- tests ---------------------------------------------------": "",
     "mocha": "mocha -r ./test/registerEsbuild",
     "test-unit": "yarn mocha \"test/unit/**/*.test.[tj]s\"",
+    "test-jesse": "DIALECT=snowflake yarn mocha \"test/unit/dialects/snowflake/query-generator.test.js\"",
```
Comment (sdepold) [inline package.json:210]: :D

File: test/integration/model/scope/merge.test.js [hunk 6]
```diff
@@ -165,6 +165,9 @@ describe(Support.getTestDialectTeaser('Model'), () => {
         const results = await Promise.all(this.scopePermutations.map(([a, b, c, d]) => this.Foo.scope(a, b, c).findOne(this.scopes[d])));
         const first = results.shift().toJSON();
         for (const result of results) {
+          // flaky test - sometimes it gets to:
+          // - bazs: [ { id: 4, quxes: [ qux7, qux8 ] }, { id: 3, quxes: [ qux5, qux6] ] } ]
+          // + bazs: [ { id: 3, quxes: [ qux5, qux6 ] }, { id: 4, quxes: [ qux7, qux8] ] } ]
```
Comment (sdepold) [inline test/integration/model/scope/merge.test.js:170]: If the order doesn't matter, let's sort it by id then

File: lib/dialects/abstract/query-generator.js [hunk 7]
```diff
@@ -914,6 +916,14 @@ class QueryGenerator {
     throw new Error(`Unknown structure passed to order / group: ${util.inspect(collection)}`);
   }
 
+  _updateQuoteIdentifier() {
```
Comment (sdepold) [inline lib/dialects/abstract/query-generator.js:919]: I'm trying to wrap my head around this... Do we really need this, or should we just rewrite the original quoteIdentifier to _quoteIdentifier in the class and move this logic to quoteIdentifier itself.

Comment (sdepold) [reply lib/dialects/abstract/query-generator.js:919]: So something like this:

Abstract
```
quoteIdentifier(identifier, force) {
  if (identifier === '*') return identifier;
  return this._quoteIdentifier(identifier, force);
}
_quoteIdentifier(identifier, force) {
  throw new Error(`quoteIdentifier for Dialect "${this.dialect}" is not implemented`);
} 
```

mysql
```
  _quoteIdentifier(identifier, force) {
     return Utils.addTicks(Utils.removeTicks(identifier, '`'), '`');
   }
```

**Review by sdepold — CHANGES_REQUESTED** [review]

File: lib/dialects/abstract/query-generator.js (hunk 1 above)
Comment (sdepold) [reply lib/dialects/abstract/query-generator.js:39]: Or remove it. See below

File: package.json (hunk 5 above)
Comment (jesse23) [reply package.json:210]: orz

File: test/integration/model/scope/merge.test.js (hunk 6 above)
Comment (jesse23) [reply test/integration/model/scope/merge.test.js:170]: I am not sure - we need to be careful between `undefined` or `default behavior desgin`

File: lib/dialects/abstract/query-generator.js (hunk 7 above)
Comment (jesse23) [reply lib/dialects/abstract/query-generator.js:919]: I have different opinion here:
- `quoteIdentifier` is an external interface 'customer/user' needs to overwrite.
-  The best approach is we use `_quoteIdentifier` in core but that will be an obvious refactor - that is why I come with this mixin like solution.

File: lib/dialects/snowflake/query-generator.js (hunk 4 above)
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:639]: sorry....orz

**Review by sdepold — APPROVED** [review]

**Comment by sdepold** [comment]
good stuff

**Comment by sdepold** [comment]
@jesse23 Seems that snowflake is somehow broken now

**Comment by jesse23** [comment]
> @jesse23 Seems that snowflake is somehow broken now

Fixed. Taking some time to test in the snowflake web UI to settle the behavior

**Pull request merged**